File: apps/ios/Voiid/Voiid/Main/VoiceNote.swift
//
//  VoiceNote.swift
//  Voiid
//
//  Hold-to-record controls and live input visualization.
//

import SwiftUI
import Combine
import AVFoundation

// MARK: - Record button (press & hold)

struct VoiceRecordButton: View {
    /// Called on release with the recorded audio bytes (.m4a) + duration (seconds).
    var onSend: (Data, TimeInterval) -> Void
    /// Told when recording starts/stops, so the composer can hand over its whole row. The
    /// recording UI CANNOT live inside this button: it needs the full width, and a 44pt
    /// capsule rendered inside a 32pt slot is what made the old one look broken.
    var onRecordingChange: (Bool) -> Void = { _ in }
    /// Live horizontal drag while recording, so the composer can draw slide-to-cancel.
    var onDrag: (CGFloat) -> Void = { _ in }
    /// Ticking duration, so the bar can show it without owning the recorder.
    var onTick: (TimeInterval) -> Void = { _ in }

    var cancelRequest: Int = 0
    var onDiscard: () -> Void = {}
    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @State private var fingerDown = false
    @State private var cancelledGesture = false
    @State private var gestureID = UUID()
    @State private var recordingError: String?
    @State private var recording = false
    @State private var seconds: TimeInterval = 0
    @State private var timer: Timer?
    @State private var recorder: AVAudioRecorder?
    @State private var fileURL: URL?
    @State private var tooShort = false
    /// Times the press-and-hold. A bare DragGesture fires on touch-down, so the "hold" is
    /// enforced here rather than by a LongPressGesture — see the gesture comment.
    @State private var holdTimer: Timer?

    var body: some View {
        // A 46pt CIRCLE, matching send — the two now SHARE one slot in the composer and
        // swap, so any size difference would show as a jump the moment you type a character.
        // (It was 32 to match the older, smaller send button; both moved to the reference's
        // 46 together.)
        //
        // It was a bare `mic.fill` glyph — no shape, no bounds — so it sat visually
        // misaligned next to the filled send button and had a vague tap target.
        Image(systemName: "mic.fill")
            .font(.system(size: 19, weight: .semibold))
            .foregroundColor(VoiidColor.primary)
            .frame(width: 46, height: 46)
            .background(VoiidColor.primary.opacity(0.12))
            .clipShape(Circle())
            .overlay(alignment: .top) {
                // "Hold to record" — the one thing a mic icon does not communicate. Shown
                // only after a too-short tap, so it teaches on failure rather than nagging.
                if tooShort {
                    Text("Hold to record")
                        .font(VoiidFont.rounded(11, .medium))
                        .foregroundColor(VoiidColor.textOnPrimary)
                        .padding(.horizontal, 10).padding(.vertical, 5)
                        .background(Capsule().fill(VoiidColor.textPrimary.opacity(0.9)))
                        .fixedSize()
                        .offset(y: -34)
                        .transition(.opacity.combined(with: .scale))
                }
            }
            .contentShape(Circle())
            // ONE DragGesture with minimumDistance 0, not a LongPress `.sequenced(before:)`.
            //
            // The sequenced form did not work: it wraps the inner drag in a
            // SequenceGesture.Value, so `.onChanged`/`.onEnded` attached to that inner drag are
            // never delivered. The drag callbacks simply never fired, which is why slide-to-
            // cancel did nothing — the bar never saw the finger move and the release always
            // took the send path.
            //
            // A bare drag fires from touch-down, so the hold is timed here instead: a 0.25s
            // timer starts recording, and a release before it fires is a tap.
            .gesture(
                DragGesture(minimumDistance: 0, coordinateSpace: .global)
                    .onChanged { value in
                        if !fingerDown {
                            fingerDown = true
                            cancelledGesture = false
                            gestureID = UUID()
                            // Touch down: arm the hold. Cancelled on an early release below.
                            armHold()
                        }
                        guard recording else { return }
                        // Only leftward travel matters; rightward is noise from the thumb
                        // rolling on the glass.
                        onDrag(min(0, value.translation.width))
                    }
                    .onEnded { value in
                        fingerDown = false
                        gestureID = UUID()
                        holdTimer?.invalidate()
                        holdTimer = nil
                        guard recording else {
                            // Released before the hold armed — a tap, not a recording.
                            if !cancelledGesture { showTooShort() }
                            return
                        }
                        finish(cancelled: value.translation.width <= RecordingBar.cancelThreshold)
                    }
            )
            .animation(.timingCurve(0.23, 1, 0.32, 1, duration: 0.15), value: tooShort)
            .scaleEffect(recording && !reduceMotion ? 1.06 : 1)
            .animation(.timingCurve(0.23, 1, 0.32, 1, duration: 0.12), value: recording)
            .accessibilityIdentifier("voice.record")
            .accessibilityLabel("Record voice message")
            .accessibilityHint("Hold to record. Slide left and release to delete.")
            .onChange(of: cancelRequest) { _, _ in
                cancelledGesture = true
                holdTimer?.invalidate()
                holdTimer = nil
                gestureID = UUID()
                if recording { finish(cancelled: true) }
            }
            .onDisappear {
                fingerDown = false
                gestureID = UUID()
                holdTimer?.invalidate()
                holdTimer = nil
                if recording { finish(cancelled: true) }
            }
            .alert("Couldn’t record", isPresented: Binding(
                get: { recordingError != nil }, set: { if !$0 { recordingError = nil } }
            )) {
                Button("OK", role: .cancel) { recordingError = nil }
            } message: { Text(recordingError ?? "") }
    }

    /// Start the hold countdown. Recording begins only if the finger is still down at 0.25s.
    private func armHold() {
        holdTimer?.invalidate()
        holdTimer = Timer.scheduledTimer(withTimeInterval: 0.25, repeats: false) { _ in
            Task { @MainActor in
                holdTimer = nil
                start()
            }
        }
    }

    /// "Hold to record" — the one thing a mic glyph cannot say. Shown on a too-short tap so it
    /// teaches on failure rather than nagging permanently.
    private func showTooShort() {
        Haptics.tap()
        withAnimation { tooShort = true }
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.6) {
            withAnimation { tooShort = false }
        }
    }

    private func start() {
        let requestID = gestureID
        AVAudioApplication.requestRecordPermission { granted in
            DispatchQueue.main.async {
                guard fingerDown, gestureID == requestID else { return }
                guard granted else {
                    recordingError = "Allow microphone access for Voiid in Settings to record voice messages."
                    return
                }
                beginRecording()
            }
        }
    }

    /// Ends the take. `cancelled` discards; otherwise it sends if long enough.
    private func finish(cancelled: Bool) {
        guard recording else { return }
        timer?.invalidate()
        timer = nil
        recording = false
        onRecordingChange(false)
        onDrag(0)
        RecordingLevel.shared.reset()
        recorder?.stop()
        let dur = seconds
        let url = fileURL
        recorder = nil
        fileURL = nil

        defer { if let url { try? FileManager.default.removeItem(at: url) } }
        try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
        guard !cancelled else { Haptics.tap(); onDiscard(); return }

        // Under half a second is a mis-tap, not a message. It used to fail SILENTLY, so the
        // user pressed the mic, nothing happened, and nothing explained why.
        guard dur >= 0.5 else { showTooShort(); return }
        guard let url, let data = try? Data(contentsOf: url) else { return }
        Haptics.success()
        onSend(data, dur)
    }

    private func beginRecording() {
        let session = AVAudioSession.sharedInstance()
        VoiceNotePlayback.pauseActive()
        do {
            try session.setCategory(.playAndRecord, mode: .default)
            try session.setActive(true)
        } catch {
            recordingError = "The microphone is unavailable. Please try again."
            return
        }
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("voiid_vn_\(UUID().uuidString).m4a")
        let settings: [String: Any] = [
            AVFormatIDKey: Int(kAudioFormatMPEG4AAC),
            AVSampleRateKey: 44100,
            AVNumberOfChannelsKey: 1,
            AVEncoderAudioQualityKey: AVAudioQuality.medium.rawValue,
        ]
        guard let rec = try? AVAudioRecorder(url: url, settings: settings) else {
            recordingError = "Couldn’t prepare the recording. Please try again."
            try? session.setActive(false, options: .notifyOthersOnDeactivation)
            return
        }
        // METERING ON: the waveform reads real input level. It used to animate random numbers,
        // which looks convincing until you realise it wiggles identically in silence — so it
        // told the user nothing about whether the mic was actually picking them up.
        rec.isMeteringEnabled = true
        guard rec.record() else {
            try? FileManager.default.removeItem(at: url)
            try? session.setActive(false, options: .notifyOthersOnDeactivation)
            recordingError = "Couldn’t start recording. Please try again."
            return
        }
        recorder = rec; fileURL = url
        Haptics.rigid()
        recording = true
        seconds = 0
        onRecordingChange(true)
        RecordingLevel.shared.reset()
        timer = Timer.scheduledTimer(withTimeInterval: 0.05, repeats: true) { _ in
            let current = rec.currentTime
            let secondChanged = Int(current) != Int(seconds)
            seconds = current
            // Only the meter needs 20 Hz. Do not rebuild the entire chat 20 times/second.
            if secondChanged { onTick(seconds) }
            rec.updateMeters()
            // averagePower is dBFS: -160 (silence) to 0 (peak). Normalised against a -50dB
            // floor, which is roughly room tone on a phone mic.
            let db = rec.averagePower(forChannel: 0)
            let level = max(0, min(1, (db + 50) / 50))
            RecordingLevel.shared.push(CGFloat(level))
        }
    }

}

// MARK: - Live input level

/// Real mic levels, published from the recorder's meter so the waveform reflects what the mic
/// is actually hearing. A shared singleton because the recorder lives in the button while the
/// waveform is drawn by the composer — they are siblings, not parent and child.
@MainActor
final class RecordingLevel: ObservableObject {
    static let shared = RecordingLevel()
    private init() {}

    /// Newest last. Fixed width so the bar scrolls rather than growing.
    @Published private(set) var levels: [CGFloat] = Array(repeating: 0.05, count: 34)

    func push(_ level: CGFloat) {
        levels.removeFirst()
        levels.append(max(0.05, level))
    }

    func reset() {
        levels = Array(repeating: 0.05, count: 34)
    }
}

/// Waveform driven by real input.
struct LiveWaveform: View {
    @ObservedObject private var source = RecordingLevel.shared
    var tint: Color = VoiidColor.primary

    var body: some View {
        Canvas { context, size in
            let count = min(source.levels.count, max(1, Int(size.width / 4.5)))
            let levels = Array(source.levels.suffix(count))
            let step = size.width / CGFloat(count)
            for (index, level) in levels.enumerated() {
                let height = max(3, size.height * level)
                let rect = CGRect(x: CGFloat(index) * step, y: (size.height - height) / 2,
                                  width: min(2.5, step), height: height)
                context.fill(Path(roundedRect: rect, cornerRadius: 1.25),
                             with: .color(tint.opacity(0.35 + 0.65 * Double(index + 1) / Double(count))))
            }
        }
        .frame(height: 22)
        .accessibilityHidden(true)
        .allowsHitTesting(false)
    }
}

// MARK: - Voice note playback bubble

struct VoiceNotePlayer: View {
    let label: String
    @State private var playing = false
    @State private var progress: CGFloat = 0
    @State private var timer: Timer?

    var body: some View {
        HStack(spacing: VoiidSpacing.sm) {
            Button {
                Haptics.tap(); togglePlay()
            } label: {
                Image(systemName: playing ? "pause.fill" : "play.fill")
                    .font(.system(size: 18)).foregroundColor(VoiidColor.primary)
            }
            // static waveform with playback progress fill
            HStack(spacing: 2) {
                ForEach(0..<22, id: \.self) { i in
                    Capsule()
                        .fill(CGFloat(i) / 22 <= progress ? VoiidColor.primary : VoiidColor.textSecondary.opacity(0.4))
                        .frame(width: 2.5, height: barHeight(i))
                }
            }
            Text(label.contains("·") ? String(label.split(separator: "·").last ?? "") : "0:03")
                .font(VoiidFont.rounded(10, .regular)).foregroundColor(VoiidColor.textSecondary)
        }
        .frame(minWidth: 180)
    }

    private func barHeight(_ i: Int) -> CGFloat {
        let pattern: [CGFloat] = [8, 14, 20, 12, 18, 10, 22, 16, 9, 15, 21]
        return pattern[i % pattern.count]
    }
    private func togglePlay() {
        playing.toggle()
        if playing {
            timer = Timer.scheduledTimer(withTimeInterval: 0.05, repeats: true) { _ in
                progress += 0.01
                if progress >= 1 { progress = 0; playing = false; timer?.invalidate() }
            }
        } else { timer?.invalidate() }
    }
}

// MARK: - Recording bar

/// The full-width bar that REPLACES the composer row while recording.
///
/// It cannot live inside the mic button. A 44pt capsule rendered inside a 32pt slot is what
/// made the old one look broken — it overflowed its container and fought the text field for
/// space. Recording is a modal state, so it takes the whole row.
///
/// SLIDE LEFT TO CANCEL is the standard gesture everywhere else and was entirely missing:
/// once you started, releasing ALWAYS sent, with no way out but sending something you did not
/// want.
struct RecordingBar: View {
    let seconds: TimeInterval
    let dragX: CGFloat
    var isDiscarding = false
    var onCancel: () -> Void
    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @ScaledMetric(relativeTo: .caption) private var captionSize = 12.0
    static let cancelThreshold: CGFloat = -90
    private var willCancel: Bool { dragX <= Self.cancelThreshold }
    private var cancelProgress: CGFloat { min(1, max(0, dragX / Self.cancelThreshold)) }

    var body: some View {
        HStack(spacing: 10) {
            Button(action: onCancel) {
                Image(systemName: isDiscarding ? "trash.fill" : "trash")
                    .symbolEffect(.bounce, value: isDiscarding && !reduceMotion)
                    .font(.system(size: 18, weight: .medium))
                    .foregroundStyle(VoiidColor.error)
                    .rotationEffect(.degrees(reduceMotion ? 0 : -12 * cancelProgress))
                    .scaleEffect(reduceMotion ? 1 : 1 + 0.1 * cancelProgress)
                    .frame(width: 44, height: 44)
                    .background(VoiidColor.error.opacity(0.08 + 0.1 * cancelProgress), in: Circle())
                    .overlay {
                        Circle().trim(from: 0, to: cancelProgress)
                            .stroke(VoiidColor.error, style: StrokeStyle(lineWidth: 2, lineCap: .round))
                            .rotationEffect(.degrees(-90))
                            .allowsHitTesting(false)
                    }
                    .contentShape(Circle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel("Delete recording")
            .accessibilityIdentifier("voice.recording.delete")

            VStack(alignment: .leading, spacing: 4) {
                HStack(spacing: 8) {
                    Circle().fill(VoiidColor.error).frame(width: 6, height: 6)
                        .accessibilityHidden(true)
                    Text(VoiceNotePlayback.time(seconds))
                        .font(VoiidFont.rounded(14, .semibold)).monospacedDigit().fixedSize()
                    LiveWaveform(tint: willCancel ? VoiidColor.error : VoiidColor.primary)
                        .frame(maxWidth: .infinity)
                }
                Text(willCancel ? "Release to delete" : "Slide left to delete · release to send")
                    .font(VoiidFont.rounded(captionSize, .medium))
                    .foregroundStyle(willCancel ? VoiidColor.error : VoiidColor.textSecondary)
                    .fixedSize(horizontal: false, vertical: true)
                    .contentTransition(.opacity)
                    .animation(.timingCurve(0.23, 1, 0.32, 1, duration: 0.15), value: willCancel)
            }
            .opacity(isDiscarding ? 0 : 1)
            .scaleEffect(isDiscarding && !reduceMotion ? 0.85 : 1, anchor: .leading)
            .offset(x: isDiscarding && !reduceMotion ? -16 : 0)
            .animation(.easeOut(duration: 0.18), value: isDiscarding)
        }
        .allowsHitTesting(!isDiscarding)
        .padding(.horizontal, 12).padding(.vertical, 4)
        .frame(minHeight: 52)
        .background(VoiidColor.fieldFill, in: RoundedRectangle(cornerRadius: 26))
        .overlay {
            RoundedRectangle(cornerRadius: 26)
                .stroke(VoiidColor.error.opacity(cancelProgress * 0.6), lineWidth: 1)
                .allowsHitTesting(false)
        }
        .onChange(of: willCancel) { _, _ in Haptics.selection() }
        .accessibilityElement(children: .contain)
        .accessibilityIdentifier("voice.recording.bar")
    }
}

File: apps/android/app/src/main/java/com/voiid/app/main/VoiceNote.kt
package com.voiid.app.main

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.slideInHorizontally
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.offset
import androidx.compose.material.icons.filled.ChevronLeft
import androidx.compose.material.icons.filled.Delete
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import kotlinx.coroutines.launch
import androidx.compose.foundation.gestures.awaitFirstDown
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Mic
import androidx.compose.material.icons.filled.Pause
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import android.media.MediaRecorder
import android.os.Build
import java.io.File
import com.voiid.app.ui.components.LocalVoiidHaptics
import com.voiid.app.ui.theme.VoiidColor
import com.voiid.app.ui.theme.VoiidFont
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive

/**
 * Voice-note recording — port of `VoiceNote.swift`.
 *
 * TWO REAL DEFECTS THIS FIXES, both of which shipped:
 *
 * 1. RELEASE ALWAYS SENT. There was no slide-to-cancel and no cancel path of any kind — the
 *    gesture watched only for `!pressed`. Once you started recording, the only way out was
 *    sending something you did not want. iOS had the same bug and fixed it; this is the port.
 *
 * 2. THE WAVEFORM WAS `Math.random()`. It looks convincing until you notice it wiggles
 *    identically in silence, so it told the user nothing about whether the mic was picking
 *    them up. It now reads `MediaRecorder.maxAmplitude`.
 */

// MARK: - Live input level

/**
 * Real mic levels, published from the recorder's meter so the waveform reflects what the mic
 * actually hears.
 *
 * A shared object because the recorder lives in the button while the waveform is drawn by the
 * composer's recording bar — they are siblings, not parent and child.
 */
object RecordingLevel {
    /** Newest last. Fixed width so the bar SCROLLS rather than growing. */
    val levels = mutableStateListOf<Float>().apply { repeat(BAR_COUNT) { add(FLOOR) } }

    fun push(level: Float) {
        levels.removeAt(0)
        levels.add(level.coerceIn(FLOOR, 1f))
    }

    fun reset() {
        for (i in levels.indices) levels[i] = FLOOR
    }

    const val BAR_COUNT = 34
    private const val FLOOR = 0.05f
}

// MARK: - Record button (press & hold)

/**
 * @param onSend           release with a long-enough take: audio bytes (.m4a) + duration.
 * @param onRecordingChange told when recording starts/stops, so the composer can hand over its
 *                         WHOLE ROW. The recording UI cannot live inside this button — a 44dp
 *                         capsule rendered inside a 32dp slot is what made the old one look
 *                         broken.
 * @param onDrag           live horizontal drag while recording, so the composer can draw
 *                         slide-to-cancel.
 * @param onTick           ticking duration, so the bar shows it without owning the recorder.
 */
@Composable
fun VoiceRecordButton(
    onSend: (ByteArray, Float) -> Unit,
    onRecordingChange: (Boolean) -> Unit = {},
    onDrag: (Float) -> Unit = {},
    onTick: (Float) -> Unit = {},
    cancelRequest: Int = 0,
    enabled: Boolean = true,
    onDiscard: () -> Unit = {},
    onError: (String) -> Unit = {},
) {
    val haptics = LocalVoiidHaptics.current
    val context = LocalContext.current
    val scope = androidx.compose.runtime.rememberCoroutineScope()
    val send by androidx.compose.runtime.rememberUpdatedState(onSend)
    val changed by androidx.compose.runtime.rememberUpdatedState(onRecordingChange)
    val dragged by androidx.compose.runtime.rememberUpdatedState(onDrag)
    val tick by androidx.compose.runtime.rememberUpdatedState(onTick)
    val discard by androidx.compose.runtime.rememberUpdatedState(onDiscard)
    val error by androidx.compose.runtime.rememberUpdatedState(onError)
    val canRecord by androidx.compose.runtime.rememberUpdatedState(enabled)
    val density = androidx.compose.ui.platform.LocalDensity.current.density
    var recording by remember { mutableStateOf(false) }
    val capture = remember { VoiceCapture(context) }
    var startedAt by remember { mutableStateOf(0L) }
    var tooShort by remember { mutableStateOf(false) }
    val permission = androidx.activity.compose.rememberLauncherForActivityResult(
        androidx.activity.result.contract.ActivityResultContracts.RequestPermission()
    ) { granted ->
        // The permission dialog ends the original touch. Require a fresh hold.
        if (granted) tooShort = true
        else error("Allow microphone access in Settings to record a voice message.")
    }
    fun finish(cancelled: Boolean) {
        if (!recording) return
        val duration = (android.os.SystemClock.elapsedRealtime() - startedAt) / 1000f
        recording = false
        val bytes = capture.finish(discard = cancelled || duration < 0.5f)
        if (cancelled) { haptics.tap(); discard() }
        changed(false)
        when {
            cancelled -> Unit
            duration < 0.5f -> tooShort = true
            bytes == null -> error("Couldn’t save this recording. Please try again.")
            else -> { haptics.success(); send(bytes, duration) }
        }
    }
    androidx.compose.runtime.LaunchedEffect(cancelRequest) {
        if (cancelRequest > 0) finish(true)
    }
    androidx.compose.runtime.LaunchedEffect(tooShort) {
        if (tooShort) { delay(1600); tooShort = false }
    }
    androidx.compose.runtime.DisposableEffect(capture) {
        onDispose { capture.finish(discard = true) }
    }
    val lifecycle = androidx.lifecycle.compose.LocalLifecycleOwner.current.lifecycle
    androidx.compose.runtime.DisposableEffect(lifecycle) {
        val observer = androidx.lifecycle.LifecycleEventObserver { _, event ->
            if (event == androidx.lifecycle.Lifecycle.Event.ON_STOP) finish(true)
        }
        lifecycle.addObserver(observer)
        onDispose { lifecycle.removeObserver(observer) }
    }
    androidx.compose.runtime.LaunchedEffect(recording) {
        if (!recording) return@LaunchedEffect
        var lastSecond = -1
        while (isActive) {
            val seconds = (android.os.SystemClock.elapsedRealtime() - startedAt) / 1000f
            if (seconds.toInt() != lastSecond) { lastSecond = seconds.toInt(); tick(seconds) }
            RecordingLevel.push(capture.level())
            delay(50)
        }
    }
    Box(
        Modifier.size(46.dp).pointerInput(Unit) {
            awaitPointerEventScope {
                while (true) {
                    val down = awaitFirstDown(requireUnconsumed = false)
                    if (!canRecord) continue
                    down.consume()
                    val pointer = down.id
                    val startX = down.position.x
                    var drag = 0f
                    var held = false
                    val holdJob = scope.launch {
                        delay(250)
                        held = true
                        if (androidx.core.content.ContextCompat.checkSelfPermission(context,
                                android.Manifest.permission.RECORD_AUDIO) != android.content.pm.PackageManager.PERMISSION_GRANTED) {
                            permission.launch(android.Manifest.permission.RECORD_AUDIO)
                        } else {
                            try {
                                capture.start()
                                startedAt = android.os.SystemClock.elapsedRealtime()
                                RecordingLevel.reset(); dragged(0f); tick(0f)
                                recording = true; changed(true); haptics.rigid()
                            } catch (_: Exception) {
                                error("Microphone unavailable. End any active call and try again.")
                            }
                        }
                    }
                    try {
                        while (true) {
                            val change = awaitPointerEvent().changes.firstOrNull { it.id == pointer } ?: break
                            drag = ((change.position.x - startX) / density).coerceAtMost(0f)
                            if (recording) dragged(drag)
                            change.consume()
                            if (!change.pressed) break
                        }
                        holdJob.cancel()
                        if (recording) finish(drag <= CANCEL_THRESHOLD_DP)
                        else if (!held) { haptics.tap(); tooShort = true }
                    } finally {
                        holdJob.cancel()
                        if (recording) finish(true)
                    }
                }
            }
        }, contentAlignment = Alignment.Center,
    ) {
        Box(Modifier.size(40.dp).clip(CircleShape).background(VoiidColor.primary.copy(alpha = 0.12f)),
            contentAlignment = Alignment.Center) {
            Icon(Icons.Default.Mic, "Hold to record voice", tint = VoiidColor.primary,
                modifier = Modifier.size(21.dp).alpha(if (recording) 0f else 1f))
        }
        AnimatedVisibility(tooShort, enter = fadeIn(), exit = fadeOut(), modifier = Modifier.offset(y = (-44).dp)) {
            Text("Hold to record", style = VoiidFont.rounded(11), color = VoiidColor.textOnPrimary,
                modifier = Modifier.clip(CircleShape).background(VoiidColor.textPrimary).padding(8.dp))
        }
    }
}

/** Owns the recorder and file together so every error/cancel path releases both. */
internal class VoiceCapture(private val context: android.content.Context) {
    private var recorder: MediaRecorder? = null
    private var file: File? = null
    fun start() {
        finish(discard = true)
        try {
            file = File.createTempFile("voice-", ".m4a", context.cacheDir)
            @Suppress("DEPRECATION")
            val r = if (Build.VERSION.SDK_INT >= 31) MediaRecorder(context) else MediaRecorder()
            recorder = r
            r.setAudioSource(MediaRecorder.AudioSource.MIC)
            r.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
            r.setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
            r.setAudioSamplingRate(44100)
            r.setAudioEncodingBitRate(64000)
            r.setOutputFile(file!!.path)
            r.prepare(); r.start()
        } catch (e: Exception) { finish(discard = true); throw e }
    }
    fun finish(discard: Boolean): ByteArray? {
        val r = recorder
        recorder = null
        return try {
            if (r == null) null else {
                r.stop()
                if (discard) null else file?.readBytes()
            }
        } catch (_: Exception) { null }
        finally {
            runCatching { r?.release() }
            file?.delete(); file = null
        }
    }
    fun level(): Float = runCatching {
        kotlin.math.sqrt((recorder?.maxAmplitude ?: 0) / 32767f).coerceIn(0f, 1f)
    }.getOrDefault(0f)
}

const val CANCEL_THRESHOLD_DP = -90f

private fun timeString(seconds: Float): String {
    val s = seconds.toInt()
    return "%d:%02d".format(s / 60, s % 60)
}

// MARK: - Live waveform while recording

/** Waveform driven by REAL input level (see [RecordingLevel]), not random numbers. */
@Composable
fun LiveWaveform(tint: androidx.compose.ui.graphics.Color = VoiidColor.primary) {
    androidx.compose.foundation.Canvas(Modifier.height(20.dp).fillMaxWidth()) {
        val levels = RecordingLevel.levels
        val step = size.width / levels.size
        levels.forEachIndexed { i, level ->
            val height = (3.dp.toPx() + (size.height - 3.dp.toPx()) * level).coerceAtMost(size.height)
            val x = step * (i + 0.5f)
            drawLine(tint.copy(alpha = 0.35f + 0.65f * i / levels.size),
                androidx.compose.ui.geometry.Offset(x, (size.height - height) / 2),
                androidx.compose.ui.geometry.Offset(x, (size.height + height) / 2),
                strokeWidth = minOf(2.5.dp.toPx(), step * 0.6f), cap = androidx.compose.ui.graphics.StrokeCap.Round)
        }
    }
}

@Composable
fun RecordingBar(seconds: Float, dragX: Float, isDiscarding: Boolean = false, onCancel: () -> Unit) {
    val willCancel = dragX <= CANCEL_THRESHOLD_DP || isDiscarding
    val tint = if (willCancel) VoiidColor.error else VoiidColor.primary
    val fade by androidx.compose.animation.core.animateFloatAsState(
        if (isDiscarding) 0f else 1f, tween(180), label = "discard")
    val rotation by androidx.compose.animation.core.animateFloatAsState(
        if (isDiscarding) -18f else 0f, tween(180), label = "trash")
    Row(Modifier.fillMaxWidth().heightIn(min = 56.dp).clip(RoundedCornerShape(28.dp))
        .background(VoiidColor.fieldFill)
        .border(1.dp, if (willCancel) tint.copy(alpha = 0.5f) else VoiidColor.fieldBorder, RoundedCornerShape(28.dp))
        .padding(horizontal = 6.dp, vertical = 4.dp),
        verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        androidx.compose.material3.IconButton(onClick = onCancel, enabled = !isDiscarding, modifier = Modifier.size(44.dp)) {
            Icon(Icons.Default.Delete, "Delete recording", tint = tint,
                modifier = Modifier.size(21.dp).graphicsLayer { rotationZ = rotation })
        }
        Text(timeString(seconds), style = VoiidFont.rounded(14, androidx.compose.ui.text.font.FontWeight.SemiBold),
            color = VoiidColor.textPrimary, maxLines = 1)
        androidx.compose.foundation.layout.Column(Modifier.weight(1f).padding(end = 12.dp)
            .graphicsLayer { alpha = fade; translationX = (1 - fade) * -16.dp.toPx() },
            verticalArrangement = Arrangement.spacedBy(3.dp)) {
            LiveWaveform(tint)
            Text(if (isDiscarding) "Recording deleted" else if (willCancel) "Release to delete" else "Slide left to delete · release to send",
                style = VoiidFont.rounded(10), color = VoiidColor.textSecondary, maxLines = 1,
                overflow = androidx.compose.ui.text.style.TextOverflow.Ellipsis)
        }
    }
}

// MARK: - Voice note playback bubble

@Composable
fun VoiceNotePlayer(label: String) {
    val haptics = LocalVoiidHaptics.current
    var playing by remember { mutableStateOf(false) }
    var progress by remember { mutableFloatStateOf(0f) }

    androidx.compose.runtime.LaunchedEffect(playing) {
        if (playing) {
            while (isActive && progress < 1f) { delay(50); progress += 0.01f }
            if (progress >= 1f) { progress = 0f; playing = false }
        }
    }

    Row(
        modifier = Modifier.widthIn(min = 180.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Box(
            modifier = Modifier.size(28.dp).clip(CircleShape).pointerInput(Unit) {
                detectTapToggle { haptics.tap(); playing = !playing }
            },
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                if (playing) Icons.Default.Pause else Icons.Default.PlayArrow, "Play",
                tint = VoiidColor.primary, modifier = Modifier.size(18.dp),
            )
        }
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(2.dp)) {
            for (i in 0 until 22) {
                val filled = i.toFloat() / 22f <= progress
                Box(
                    Modifier
                        .width(2.5.dp)
                        .height(barHeight(i).dp)
                        .clip(CircleShape)
                        .background(if (filled) VoiidColor.primary else VoiidColor.textSecondary.copy(alpha = 0.4f)),
                )
            }
        }
        Text(
            if (label.contains("·")) label.substringAfterLast("·").trim() else "0:03",
            style = VoiidFont.rounded(10), color = VoiidColor.textSecondary,
        )
    }
}

private fun barHeight(i: Int): Int {
    val pattern = intArrayOf(8, 14, 20, 12, 18, 10, 22, 16, 9, 15, 21)
    return pattern[i % pattern.size]
}

// simple tap detector used by the play button
private suspend fun androidx.compose.ui.input.pointer.PointerInputScope.detectTapToggle(onTap: () -> Unit) {
    awaitPointerEventScope {
        while (true) {
            awaitFirstDown(requireUnconsumed = false)
            var released = false
            while (!released) {
                val e = awaitPointerEvent()
                if (e.changes.all { !it.pressed }) released = true
            }
            onTap()
        }
    }
}
